src/input.rs

```diff
@@ -84,9 +88,17 @@ impl<'a> InputKind<'a> {
     }
 }
 
+#[derive(Clone)]
+pub(crate) struct InputPermissions {
+    pub(crate) mode: u32,
+}
+
 #[derive(Clone, Default)]
 pub(crate) struct InputMetadata {
     pub(crate) user_provided_name: Option<PathBuf>,
+    pub(crate) size: Option<u64>,
+    pub(crate) permissions: Option<InputPermissions>,
```

Comment:
This might be an irrelevant point if we go with option 1, but why not store the `mode` directly? I suspect you introduced `InputPermissions` as a form of future-proofed abstraction, but since this is internal to `bat` and not part of the API I would prefer if we kept it as simple as possible.